### PR TITLE
Fix: ServiceBinding resources are not deployed with odo deploy

### LINKED ISSUE #5883: ServiceBinding resources are not deployed with `odo deploy`
/kind bug

## What versions of software are you using?

```
$ odo version
odo v3.0.0-alpha3 (8c99ce920)
```

## How did you run odo exactly?

Apply this cluster resource:
```
apiVersion: postgresql.k8s.enterprisedb.io/v1
kind: Cluster
metadata:
  name: cluster-sample
spec:
  instances: 1
  logLevel: info
  primaryUpdateStrategy: unsupervised
  storage:
    size: 1Gi
```

```
variables:
  CONTAINER_IMAGE: quay.io/tkral/devfile-nodejs-deploy:latest
commands:
- exec:
    commandLine: GOCACHE=${PROJECT_SOURCE}/.cache go build main.go
    component: runtime
    group:
      isDefault: true
      kind: build
    hotReloadCapable: false
    workingDir: ${PROJECT_SOURCE}
  id: build
- exec:
    commandLine: ./main
    component: runtime
    group:
      isDefault: true
      kind: run
    hotReloadCapable: false
    workingDir: ${PROJECT_SOURCE}
  id: run

- id: deploy
  composite:
    commands:
    - build-image
    - deploy-deployment
    - deploy-service
    - deploy-binding
    group:
      kind: deploy
      isDefault: true
- id: build-image
  apply:
    component: prod-image
- id: deploy-deployment
  apply:
    component: outerloop-deploy
- id: deploy-service
  apply:
    component: outerloop-service
- id: deploy-binding
  apply:
    component: my-go-app-cluster-sample

components:
- container:
    dedicatedPod: false
    endpoints:
    - name: http
      secure: false
      targetPort: 8080
    image: quay.io/devfile/golang:latest
    memoryLimit: 1024Mi
    mountSources: true
  name: runtime

- name: prod-image
  image:
    imageName: "{{CONTAINER_IMAGE}}"
    dockerfile:
      uri: ./Dockerfile
      buildContext: ${PROJECT_SOURCE}
- name: outerloop-deploy
  kubernetes:
    inlined: |
      kind: Deployment
      apiVersion: apps/v1
      metadata:
        name: devfile-deploy
      spec:
        replicas: 1
        selector:
          matchLabels:
            app: devfile-deploy
        template:
          metadata:
            labels:
              app: devfile-deploy
          spec:
            containers:
              - name: main
                image: "{{CONTAINER_IMAGE}}"
                resources: {}
- name: outerloop-service
  kubernetes:
    inlined: |
        apiVersion: v1
        kind: Service
        metadata:
          labels:
            app: devfile-deploy
          name: devfile-deploy
        spec:
          ports:
          - name: http-3000
            port: 3000
            protocol: TCP
            targetPort: 3000
          selector:
            app: devfile-deploy
          type: LoadBalancer

- kubernetes:
    inlined: |
      apiVersion: binding.operators.coreos.com/v1alpha1
      kind: ServiceBinding
      metadata:
        name: my-go-app-cluster-sample
      spec:
        application:
          group: apps
          kind: Deployment
          name: devfile-deploy
          version: v1
        bindAsFiles: true
        detectBindingResources: true
        services:
        - group: postgresql.k8s.enterprisedb.io
          id: my-go-app-cluster-sample
          kind: Cluster
          name: cluster-sample
          resource: clusters
          version: v1
      status:
        secret: ""
  name: my-go-app-cluster-sample
metadata:
  description: Stack with the latest Go version
  displayName: Go Runtime
  icon: https://raw.githubusercontent.com/devfile-samples/devfile-stack-icons/main/golang.svg
  language: go
  name: my-go-app
  projectType: go
  tags:
  - Go
  version: 1.0.0
schemaVersion: 2.2.0
starterProjects:
- git:
    checkoutFrom:
      revision: main
    remotes:
      origin: https://github.com/devfile-samples/devfile-stack-go.git
  name: go-starter

```

```
$ PODMAN_CMD=echo odo deploy
  __
 /  \__     Deploying the application using my-go-app Devfile
 \__/  \    Namespace: prj2
 /  \__/    odo version: v3.0.0-beta2
 \__/

↪ Building & Pushing Container: quay.io/tkral/devfile-nodejs-deploy:latest
 •  Building image locally  ...
build -t quay.io/tkral/devfile-nodejs-deploy:latest -f /home/phmartin/Documents/tests/5883/Dockerfile /home/phmartin/Documents/tests/5883
 ✓  Building image locally 
 •  Pushing image to container registry  ...
push quay.io/tkral/devfile-nodejs-deploy:latest
 ✓  Pushing image to container registry 

↪ Deploying Kubernetes Component: devfile-deploy
 ✓  Creating kind Deployment [15ms]

↪ Deploying Kubernetes Component: devfile-deploy
 ✓  Creating kind Service [13ms]

↪ Deploying Kubernetes Component: my-go-app-cluster-sample

Your Devfile has been successfully deployed
```

## Actual behavior

```
$ kubectl get servicebindings.binding.operators.coreos.com 
No resources found in demo namespace.
```

## Expected behavior

The service binding must be created in the cluster

## Any logs, error output, etc?

The problem occurs because the `odo dev` command has a special treatment for ServiceBinding resources, and `odo deploy` is reusing the same function (see `service.PushKubernetesResource`)

### PR DESCRIPTION
**Pull request by valaparthvi**
Signed-off-by: Parthvi Vala <pvala@redhat.com>

**What type of PR is this:**
/kind bug

**What does this PR do / why we need it:**
ServiceBinding resources are now allowed to be deployed on the cluster with odo deploy.

**Which issue(s) this PR fixes:**

Fixes #5883

**PR acceptance criteria:**

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**
1.
```sh
odo init --devfile go --name my-go-app --starter go-starter
``` 
2. Copy the contents of the devfile.yaml as mentioned in #5883 
3. Deploy the bindable kind instance as mentioned in #5883 
4. Run `odo deploy`.
```sh
PODMAN_CMD=echo odo deploy
```

Note: Test case for `odo dev` without SBO is pending.

### CODE REVIEW

**Comment by feloy** [comment]
The change is small, but complexifies the logic of the code: 
- For `odo dev`, we are using PushLinks to create the bindings with or without operators
- For `odo deploy`, we are directly applying the servicebinding resource

I think that making some refactoring around the functions `PushLinks`, `pushDevfileKubernetesComponents`, `PushKubernetesResources` and `PushKubernetesResource` would be helpful, so the code would be usable for `odo dev` and `odo deploy` in a nice way.

**Comment by valaparthvi** [comment]
@feloy There were a few complications with the refactoring when SBO is not installed. Since we're using library, it is required that the deployment in question be present before creating the secret. Because of this, we'll create the link after the component deployment is created if SBO is not present, otherwise it'll be created normally like any other K8s component.

With `odo deploy`, we don't really need to deal with the Without SBO scenario.

**Comment by feloy** [comment]
> @feloy There were a few complications with the refactoring when SBO is not installed. Since we're using library, it is required that the deployment in question be present before creating the secret. Because of this, we'll create the link after the component deployment is created if SBO is not present, otherwise it'll be created normally like any other K8s component.
> 
> With `odo deploy`, we don't really need to deal with the Without SBO scenario.

@valaparthvi Did you try by creating all Kubernetes resources (servicebindings and all other resources) after the deployment?

**Comment by valaparthvi** [comment]
> > @feloy There were a few complications with the refactoring when SBO is not installed. Since we're using library, it is required that the deployment in question be present before creating the secret. Because of this, we'll create the link after the component deployment is created if SBO is not present, otherwise it'll be created normally like any other K8s component.
> > With `odo deploy`, we don't really need to deal with the Without SBO scenario.
> 
> @valaparthvi Did you try by creating all Kubernetes resources (servicebindings and all other resources) after the deployment?

Theoretically, I think that would work, but then we also talked about creating all the resources beforehand to avoid restarting the pod and speeding up the process. So I didn't go that way.

**Comment by feloy** [comment]
> > > @feloy There were a few complications with the refactoring when SBO is not installed. Since we're using library, it is required that the deployment in question be present before creating the secret. Because of this, we'll create the link after the component deployment is created if SBO is not present, otherwise it'll be created normally like any other K8s component.
> > > With `odo deploy`, we don't really need to deal with the Without SBO scenario.
> > 
> > 
> > @valaparthvi Did you try by creating all Kubernetes resources (servicebindings and all other resources) after the deployment?
> 
> Theoretically, I think that would work, but then we also talked about creating all the resources beforehand to avoid restarting the pod and speeding up the process. So I didn't go that way.

Thinking again about this, tt's matter of milliseconds. We would create Deployment and SBs very quickly, we are not 100% sure in which order they will be handled by controllers on the cluster.

And creating the SB before the Deployment can raise an error for the SBO because the Deployment does not exist when the SB is created, and I'm not sure how the error is recovered (either quickly or not).

**Comment by valaparthvi** [comment]
>And creating the SB before the Deployment can raise an error for the SBO because the Deployment does not exist when the SB is created, and I'm not sure how the error is recovered (either quickly or not).

It seems to be working fine so far. Creating SB before Deployment has not been a problem so far.

**Comment by valaparthvi** [comment]
>Thinking again about this, it's a matter of milliseconds. We would create Deployment and SBs very quickly, we are not 100% sure in which order they will be handled by controllers on the cluster.

Alright, let's go with creating K8s resources are the core deployment is created, it seems logical.

**Comment by feloy** [comment]
> > Thinking again about this, it's a matter of milliseconds. We would create Deployment and SBs very quickly, we are not 100% sure in which order they will be handled by controllers on the cluster.
> 
> Alright, let's go with creating K8s resources are the core deployment is created, it seems logical.

I'm trying to explore this direction, as it seems the most generic one, so we don't have to handle specific cases, and we can "hide" the fact that we are using the SBO library at the lowest possible level (in the kclient package).

**Comment by valaparthvi** [comment]
> > > Thinking again about this, it's a matter of milliseconds. We would create Deployment and SBs very quickly, we are not 100% sure in which order they will be handled by controllers on the cluster.
> > 
> > 
> > Alright, let's go with creating K8s resources are the core deployment is created, it seems logical.
> 
> I'm trying to explore this direction, as it seems the most generic one, so we don't have to handle specific cases at high levels, and we can "hide" the fact that we are using the SBO library at the lowest possible level (in the kclient package).

I find it difficult to imagine handling this case in the kclient package.

**Comment by valaparthvi** [comment]
> > > > > Thinking again about this, it's a matter of milliseconds. We would create Deployment and SBs very quickly, we are not 100% sure in which order they will be handled by controllers on the cluster.
> > > > 
> > > > 
> > > > Alright, let's go with creating K8s resources are the core deployment is created, it seems logical.
> > > 
> > > 
> > > I'm trying to explore this direction, as it seems the most generic one, so we don't have to handle specific cases at high levels, and we can "hide" the fact that we are using the SBO library at the lowest possible level (in the kclient package).
> > 
> > 
> > I find it difficult to imagine handling this case in the kclient package.
> 
> Or in the `pkg/service` package?

Yes, I was thinking that too.

File: pkg/service/service.go [hunk 1]
```diff
@@ -241,15 +241,17 @@ func PushKubernetesResources(client kclient.ClientInterface, devfileObj parser.D
...
+	sboSupported, err := client.IsServiceBindingSupported()
+	if err != nil {
+		return err
 	}
 
-	isOp, err := isOperatorBackedService(client, u)
-	if err != nil {
-		return false, err
+	// If the component is of Kind: ServiceBinding, trying to run in Dev mode and SBO is not installed, run it without operator.
+	if isLinkResource(u.GetKind()) && mode == odolabels.ComponentDevMode && !sboSupported {
```
Comment (feloy) [inline pkg/service/service.go:251]: Why only for dev mode? Don't we want to use the library for `odo deploy`?

**Review by feloy — COMMENTED** [review]
LGTM

One question though

File: pkg/service/service.go (hunk 1 above)
Comment (valaparthvi) [reply pkg/service/service.go:251]: With `odo deploy`, I would expect user to know that they're supposed to install ServiceBindingOperator if they intend to use ServiceBinding, just like they would with any other CRD.
Thinking about it this way, I think it makes sense to only check for `odo dev`, but if we want to provide consistency, having the same behavior for `odo deploy` makes sense.

Comment (feloy) [reply pkg/service/service.go:251]: OK.

I can see some tests failing about odo logs, probably not related, I'll restart the tests to be sure

**Comment by valaparthvi** [comment]
The failures are real, I'm debugging them right now.

File: pkg/service/service.go (hunk 1 above)
Comment (valaparthvi) [reply pkg/service/service.go:251]: No, they're failing locally as well. I am debugging them.

File: pkg/devfile/adapters/kubernetes/component/adapter.go [hunk 2]
```diff
@@ -132,6 +127,12 @@ func (a Adapter) Push(parameters adapters.PushParameters, componentStatus *watch
 		return fmt.Errorf("unable to create or update component: %w", err)
 	}
 
+	// Create all the K8s components defined in the devfile
+	k8sComponents, err := a.pushDevfileKubernetesComponents(labels, odolabels.ComponentDevMode)
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/devfile/adapters/kubernetes/component/adapter.go:131]: Yes you are right, you could set the owneerref when you're creating them

**Comment by valaparthvi** [comment]
OpenShift test failure is probably flake or it failed because of resource crunch. The test passes locally.

```sh
odo delete command tests
/go/odo_1/tests/integration/cmd_delete_test.go:13
  when deleting a component containing preStop event that is deployed with DEV [BeforeEach]
  /go/odo_1/tests/integration/cmd_delete_test.go:257
    should contain preStop events list
    /go/odo_1/tests/integration/cmd_delete_test.go:270
...
...
...
  Timed out after 180.001s.
  Expected
      <string>:   __
       /  \__     Developing using the nodejs Devfile
       \__/  \    Namespace: cmd-delete-test270afe
       /  \__/    odo version: v3.0.0-beta3
       \__/
      
      â†ª Deploying to the cluster in developer mode
       â€¢  Waiting for Kubernetes resources  ...
       âš   Pod is Pending
       âš   0/3 nodes are available: 3 Insufficient memory.
       âš   0/3 nodes are available: 3 Insufficient memory.
       âš   0/3 nodes are available: 3 Insufficient memory.
      
  to contain substring
      <string>: Press Ctrl+c to exit
```

**Comment by feloy** [comment]
> ```shell
> Insufficient memory
> ```

> Insufficient memory

This error is probably due to old pods from old tests running on the cluster. Could you connect to the Openshift cluster and check if there are old tests bamespaces ? I have also seen tests pods in the default namespace this week, not sure where they were coming from.

**Comment by feloy** [comment]
/lgtm
The code looks good 
I'll try to make more tests before to approve

**Comment by feloy** [comment]
/approve

/override ci/prow/unit-and-validate-test
/override ci/prow/v4.10-integration-e2e
Infra errors
Tests pass on IBM Cloud

**Pull request merged**